sessionctx/variable/session.go

```diff
@@ -805,3 +813,73 @@ type BatchSize struct {
 	// MaxChunkSize defines max row count of a Chunk during query execution.
 	MaxChunkSize int
 }
+
+const (
+	// SlowLogPrefixStr is slow log row prefix.
+	SlowLogPrefixStr = "# "
+	// SlowLogSpaceMarkStr is slow log space mark.
+	SlowLogSpaceMarkStr = ": "
+	// SlowLogSQLSuffixStr is slow log suffix.
+	SlowLogSQLSuffixStr = ";"
+	// SlowLogTimeStr is slow log field name.
+	SlowLogTimeStr = "Time"
+	// SlowLogStartPrefixStr is slow log start row prefix.
+	SlowLogStartPrefixStr = SlowLogPrefixStr + SlowLogTimeStr + SlowLogSpaceMarkStr
+	// SlowLogTxnStartTSStr is slow log field name.
+	SlowLogTxnStartTSStr = "Txn_start_ts"
+	// SlowLogUserStr is slow log field name.
+	SlowLogUserStr = "User"
+	// SlowLogConnIDStr is slow log field name.
+	SlowLogConnIDStr = "Conn_ID"
+	// SlowLogQueryTimeStr is slow log field name.
+	SlowLogQueryTimeStr = "Query_time"
+	// SlowLogDBStr is slow log field name.
+	SlowLogDBStr = "DB"
+	// SlowLogIsInternalStr is slow log field name.
+	SlowLogIsInternalStr = "Is_internal"
+	// SlowLogIndexNamesStr is slow log field name.
+	SlowLogIndexNamesStr = "Index_names"
+	// SlowLogQuerySQLStr is slow log field name.
+	SlowLogQuerySQLStr = "Query" // use for slow log table, slow log will not print this field name but print sql directly.
+)
+
+// SlowLogFormat uses for formatting slow log.
+// The slow log output is like below:
+// # Time: 2019-02-12-19:33:56.571953 +0800
+// # Txn_start_ts: 406315658548871171
+// # User: root@127.0.0.1
+// # Conn_ID: 6
+// # Query_time: 4.895492
+// # Process_time: 0.161 Request_count: 1 Total_keys: 100001 Processed_keys: 100000
+// # DB: test
+// # Is_internal: false
+// select * from t_slim;
+func (s *SessionVars) SlowLogFormat(txnTS uint64, costTime time.Duration, execDetail execdetails.ExecDetails, indexIDs string, sql string) string {
+	var buf bytes.Buffer
+	execDetailStr := execDetail.String()
+	buf.WriteString(SlowLogPrefixStr + SlowLogTxnStartTSStr + SlowLogSpaceMarkStr + strconv.FormatUint(txnTS, 10) + "\n")
+	if s.User != nil {
+		buf.WriteString(SlowLogPrefixStr + SlowLogUserStr + SlowLogSpaceMarkStr + s.User.String() + "\n")
+	}
+	if s.ConnectionID != 0 {
+		buf.WriteString(SlowLogPrefixStr + SlowLogConnIDStr + SlowLogSpaceMarkStr + strconv.FormatUint(s.ConnectionID, 10) + "\n")
+	}
+	buf.WriteString(SlowLogPrefixStr + SlowLogQueryTimeStr + SlowLogSpaceMarkStr + strconv.FormatFloat(costTime.Seconds(), 'f', -1, 64) + "\n")
+	if len(execDetailStr) > 0 {
+		buf.WriteString(SlowLogPrefixStr + execDetailStr + "\n")
+	}
+	if len(s.CurrentDB) > 0 {
+		buf.WriteString(SlowLogPrefixStr + SlowLogDBStr + SlowLogSpaceMarkStr + s.CurrentDB + "\n")
+	}
+	if len(indexIDs) > 0 {
+		buf.WriteString(SlowLogPrefixStr + SlowLogIndexNamesStr + SlowLogSpaceMarkStr + indexIDs + "\n")
+	}
+	buf.WriteString(SlowLogPrefixStr + SlowLogIsInternalStr + SlowLogSpaceMarkStr + strconv.FormatBool(s.InRestrictedSQL) + "\n")
+	if len(sql) > 0 {
```

Comment:
Under what circumstances will encounter the situation of `len(sql) == 0 `?

Reply:
Normally It won't be `len(sql) = 0`, but just in case, we should also print ";" when `len(sql)=0`. Parser need ";" to indicate the end of slow log.